gensim/similarities/termsim.py

```diff
@@ -315,22 +401,37 @@ def inner_product(self, X, Y, normalized=False):
 
             result = X.T.dot(matrix).dot(Y)
 
-            if normalized:
+            norm = 1.0
+
+            if normalized_X:
                 X_norm = X.T.dot(matrix).dot(X)[0, 0]
+                assert X_norm >= 0.0, non_negative_norm_assertion_message
+                if normalized_X == 'maintain' and X_norm > 0.0:
+                    X_norm /= X.T.dot(X)
+                X_norm = sqrt(X_norm)
+                if X_norm > 0.0:
+                    norm *= X_norm
+
+            if normalized_Y:
                 Y_norm = Y.T.dot(matrix).dot(Y)[0, 0]
+                assert Y_norm >= 0.0, non_negative_norm_assertion_message
+                if normalized_Y == 'maintain' and Y_norm > 0.0:
+                    Y_norm /= Y.T.dot(Y)
+                Y_norm = sqrt(Y_norm)
+                if Y_norm > 0.0:
+                    norm *= Y_norm
 
-                assert \
-                    X_norm > 0.0 and Y_norm > 0.0, \
-                    u"sparse documents must not contain any explicit zero entries and the similarity matrix S " \
-                    u"must satisfy x^T * S * x > 0 for any nonzero bag-of-words vector x."
+            if normalized_X or normalized_Y:
+                result *= 1.0 / norm
 
-                result /= sqrt(X_norm) * sqrt(Y_norm)
+            if normalized_X is True and normalized_Y is True:
                 result = np.clip(result, -1.0, 1.0)
 
             return result[0, 0]
         elif not is_corpus_X or not is_corpus_Y:
             if is_corpus_X and not is_corpus_Y:
-                is_corpus_X, X, is_corpus_Y, Y = is_corpus_Y, Y, is_corpus_X, X  # make Y the corpus
+                is_corpus_X, X, is_corpus_Y, Y, normalized_X, normalized_Y = \
```

Comment:
That line break is ugly, I think multi-line assignments would be more readable here.

Reply:
Fixed in 5d68e131.